Duplicati/Library/Main/Database/LocalDeleteDatabase.cs

```diff
@@ -109,18 +109,24 @@ private void InitializeCommands()
             }
         }
 
-        public IEnumerable<IListResultFileset> Filesets
+        /// <summary>
+        /// Returns a collection of IListResultFilesets, where the Version is the backup version number
+        /// exposed to the user.  This is in contrast to other cases where the Version is the ID in the
+        /// Fileset table.
+        /// </summary>
+        internal IEnumerable<IListResultFileset> FilesetsWithBackupVersion
         {
             get
             {
                 List<IListResultFileset> filesets = new List<IListResultFileset>();
                 using (IDbCommand cmd = this.m_connection.CreateCommand())
                 {
-                    using (IDataReader reader = cmd.ExecuteReader(@"SELECT ""ID"", ""IsFullBackup"", ""Timestamp"" FROM ""Fileset"" ORDER BY ""Timestamp"" DESC "))
+                    using (IDataReader reader = cmd.ExecuteReader(@"SELECT ROW_NUMBER() OVER (ORDER BY ""Timestamp"" DESC) Version, ""IsFullBackup"", ""Timestamp"" FROM ""Fileset"""))

```

Comment:
I'm not sure why the `PurgeBrokenFilesTest` is still failing.  The test passes for me when I run the tests locally.

Could there be an issue with using `ROW_NUMBER()` on Travis?  I think we are using the same versions of `System.Data.SQLite` (1.0.111).

Reply:
What happens if you run the test with the pipeline/start.sh script? Because this should be travis independent.

Reply:
I haven't used the `pipeline/start.sh` script, but the test passes for me in both MonoDevelop and Rider.  I noticed that Travis appears to be using mono 6.6.0.161, whereas I am using mono 6.4.0 locally.  Could that be an issue?

Reply:
Maybe. As a test you could change the docker image to be used to mono 6.4. Should be changed in stage_build/jobs.sh iirc.

Reply:
I tested this with mono 6.4 in the docker image and the test still fails.  Looking at the nuget packages folder in the `/tmp/.duplicati_build` and `/tmp/.duplicati_test` folders, they match what is in my local development folder.

Are you able to see if the `PurgeBrokenFilesTest` in this branch is passing in Rider for you?

Reply:
This is really strange.  When I run `pipeline/start.sh`, I see the test failure:
```
Dec 14 21:14:56 1) Error : Duplicati.UnitTest.PurgeTesting.PurgeBrokenFilesTest
Dec 14 21:14:56 Mono.Data.Sqlite.SqliteException : SQLite error
Dec 14 21:14:56 near "(": syntax error
Dec 14 21:14:56   at Mono.Data.Sqlite.SQLite3.Prepare (Mono.Data.Sqlite.SqliteConnection cnn, System.String strSql, Mono.Data.Sqlite.SqliteStatement previous, System.UInt32 timeoutMS, System.String& strRemain) [0x001f8] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteCommand.BuildNextCommand () [0x000d3] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteCommand.GetStatement (System.Int32 index) [0x00008] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at (wrapper remoting-invoke-with-check) Mono.Data.Sqlite.SqliteCommand.GetStatement(int)
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteDataReader.NextResult () [0x000b3] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteDataReader..ctor (Mono.Data.Sqlite.SqliteCommand cmd, System.Data.CommandBehavior behave) [0x0004e] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at (wrapper remoting-invoke-with-check) Mono.Data.Sqlite.SqliteDataReader..ctor(Mono.Data.Sqlite.SqliteCommand,System.Data.CommandBehavior)
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x00006] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at Mono.Data.Sqlite.SqliteCommand.ExecuteDbDataReader (System.Data.CommandBehavior behavior) [0x00000] in <1c97a93cf49e412e9e991ba4fa395bc9>:0 
Dec 14 21:14:56   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader () [0x00000] in <ae96b2e871e140cf907c5d70e1fdda5a>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Database.ExtensionMethods.ExecuteReader (System.Data.IDbCommand self, System.Boolean writeLog, System.String cmd, System.Object[] values) [0x0005e] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Database.ExtensionMethods.ExecuteReader (System.Data.IDbCommand self, System.String cmd, System.Object[] values) [0x00000] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Database.LocalDeleteDatabase.get_FilesetsWithBackupVersion () [0x0001d] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Operation.DeleteHandler.DoRun (Duplicati.Library.Main.Database.LocalDeleteDatabase db, System.Data.IDbTransaction& transaction, System.Boolean hasVerifiedBacked, System.Boolean forceCompact, Duplicati.Library.Main.BackendManager sharedManager) [0x00056] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Operation.PurgeBrokenFilesHandler.Run (Duplicati.Library.Utility.IFilter filter) [0x0039d] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Controller+<>c__DisplayClass35_0.<PurgeBrokenFiles>b__0 (Duplicati.Library.Main.PurgeBrokenFilesResults result) [0x0001c] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Controller.RunAction[T] (T result, System.String[]& paths, Duplicati.Library.Utility.IFilter& filter, System.Action`1[T] method) [0x0026f] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Controller.RunAction[T] (T result, System.Action`1[T] method) [0x00009] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.Library.Main.Controller.PurgeBrokenFiles (Duplicati.Library.Utility.IFilter filter) [0x0001a] in <3ae981f3f23d409abaef5211e2158058>:0 
Dec 14 21:14:56   at Duplicati.UnitTest.PurgeTesting.PurgeBrokenFilesTest () [0x0044f] in <1c9d8eaba09e43bba2809428372a5926>:0 
Dec 14 21:14:56   at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
Dec 14 21:14:56   at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <d2ec5c92492f4d6ba8c422bdf574b786>:0 
Dec 14 21:14:56 
Dec 14 21:14:56 Run Settings
Dec 14 21:14:56     DisposeRunners: True
Dec 14 21:14:56     WorkDirectory: /application
Dec 14 21:14:56     NumberOfTestWorkers: 1
Dec 14 21:14:56     ImageRuntimeVersion: 4.0.30319
Dec 14 21:14:56     ImageTargetFrameworkName: .NETFramework,Version=v4.6.2
Dec 14 21:14:56     ImageRequiresX86: False
Dec 14 21:14:56     ImageRequiresDefaultAppDomainAssemblyResolver: False
Dec 14 21:14:56 
Dec 14 21:14:56 Test Run Summary
Dec 14 21:14:56   Overall result: Failed
Dec 14 21:14:56   Test Count: 2, Passed: 1, Failed: 1, Warnings: 0, Inconclusive: 0, Skipped: 0
Dec 14 21:14:56     Failed Tests - Failures: 0, Errors: 1, Invalid: 0
Dec 14 21:14:56   Start time: 2019-12-15 05:14:07Z
Dec 14 21:14:56     End time: 2019-12-15 05:14:56Z
Dec 14 21:14:56     Duration: 49.765 seconds
```
If I then `cd` to `/tmp/.duplicati_test` and run
```
nunit3-console Duplicati/UnitTest/bin/Release/Duplicati.UnitTest.dll --where:cat==Purge
```
the tests pass:
```
Run Settings
    DisposeRunners: True
    WorkDirectory: /tmp/.duplicati_test
    NumberOfTestWorkers: 1
    ImageRuntimeVersion: 4.0.30319
    ImageTargetFrameworkName: .NETFramework,Version=v4.6.2
    ImageRequiresX86: False
    ImageRequiresDefaultAppDomainAssemblyResolver: False

Test Run Summary
  Overall result: Passed
  Test Count: 2, Passed: 2, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2019-12-15 05:17:55Z
    End time: 2019-12-15 05:18:48Z
    Duration: 52.673 seconds
```

Reply:
I don't quite understand how the sqlite dependencies work.  Do we depend on system sqlite libraries for runtime?  On my machine, I have sqlite 3.30, but Debian stretch has [sqlite 3.16.2](https://packages.debian.org/stretch/sqlite3), which doesn't support the `ROW_NUMBER` window function.  We are compiling with the sqlite 1.0.111 nuget package, which supports sqlite 3.28.  This might explain the different behavior.

Reply:
Good question, I can change the docker config or add in the setup of the container a newer version of sqlite, but if we do not depend on system libraries, then we should use the included libs.

Reply:
I'm not certain, but I have a feeling that in some cases we do fall back to the system libraries.  This concerns me, since several distributions (e.g., Debian) ship with outdated versions of sqlite.  I think the ideal case is to to ship the correct libraries and reduce the dependency on system libraries.  There's also the `Microsoft.Data.Sqlite` package, but I'm not sure how that differs from the `System.Data.Sqlite` and `Mono.Data.Sqlite` ones.

Perhaps the best approach is to not use `ROW_NUMBER` here, and deal with our sqlite dependencies as a separate task?

Reply:
Reivision 610fa4ffeae90640c50f6dd801810b433add9fee removes the use of the `ROW_NUMBER` window function and generates the backup version numbers separately.  I expect that the tests should pass now.

I also created issue #4024 to remind us to solidify the SQLite dependencies.